metadata_layer.py

```diff
@@ -502,6 +504,128 @@ def __init__(self, content_type, drm_scheme, link=None, rights_uri=None):
+
+    NO_VALUE = object()
+
+    def __init__(self, start=NO_VALUE, finish=NO_VALUE, achievements=NO_VALUE,
+                 counter=NO_VALUE, exception=NO_VALUE):
+        """A constructor intended to be used by a service to customize its
+        eventual Timestamp.
+
+        service, service_type, and collection cannot be set through
+        this constructor, because they are generally not under the
+        control of the code that runs the service. They are set
+        afterwards, in finalize().
+
+        :param start: The time that the service should be considered to
+           have started running.
+        :param finish: The time that the service should be considered
+           to have stopped running.
+        :param achievements: A string describing what was achieved by the
+           service.
+        :param counter: A single integer item of state representing the point
+           at which the service left off.
+        :param exception: A traceback representing an exception that stopped
+           the progress of the service.
+        """
+
+        # These are set by finalize().
+        self.service = self.NO_VALUE
+        self.service_type = self.NO_VALUE
+        self.collection_id = self.NO_VALUE
+
+        self.start = start
+        self.finish = finish
+        self.achievements = achievements
+        self.counter = counter
+        self.exception = exception
+
+    @property
+    def is_failure(self):
+        """Does this TimestampData represent an unrecoverable failure?"""
+        return self.exception not in (None, self.NO_VALUE)
+
+    @property
+    def is_complete(self):
+        """Does this TimestampData represent an operation that has
+        completed?
+
+        An operation is completed if it has failed, or if the time of its
+        completion is known.
+        """
+        return self.is_failure or self.finish not in (None, self.NO_VALUE)
+
+    def finalize(self, service, service_type, collection, start=NO_VALUE,
+                 finish=NO_VALUE, achievements=NO_VALUE, counter=NO_VALUE,
+                 exception=NO_VALUE):
+        """Finalize any values that were not set during the constructor.
+
+        This is intended to be run by the code that originally ran the
+        service.
+
+        The given values for `start`, `finish`, `achievements`,
+        `counter`, and `exception` will be used only if the service
+        did not specify its own values for those fields.
+        """
+        self.service = service
+        self.service_type = service_type
+        if collection is None:
+            self.collection_id = None
+        else:
+            self.collection_id = collection.id
+        if self.start is self.NO_VALUE:
+            self.start = start
+        if self.finish is self.NO_VALUE:
+            if finish is self.NO_VALUE:
+                finish = datetime.datetime.utcnow()
+            self.finish = finish
+        if self.start is self.NO_VALUE:
```

Comment:
This check is being done above. This is the case where both instances of `start` are `NO_VALUE`? And why do you need to set it to `self.finish`?

Reply:
Yes, exactly -- neither the constructor nor `finalize` explained  when the service started running. In that case we use the only date we have available -- the time when the service started running. If it's also the case that no one explained when the service _stopped_ running (which is pretty likely) then we use the current time for both dates.

The goal is just to have both dates set to some value for all `Timestamp` objects, so that we can do things like sort by one field or another in database queries.